File: 06-breadth-first-search/python/seller_search.py
from collections import deque

def person_is_seller(person):
    return person[-1] == 'm'

def seller_search(graph, name):
    """Implementation of breadth-first search to find a mango seller from a certain node"""
    search_queue = deque(graph[name])
    searched = {}

    while search_queue:
        person = search_queue.popleft()
        if not searched.get(person):
            if person_is_seller(person):
                return (True, person)
            else:
                search_queue += graph[person]
                searched[person] = True
    return (False, None)

File: 06-breadth-first-search/ruby/seller_search.rb
# frozen_string_literal: true

def person_is_seller?(person)
  person[-1] == 'm'
end

def seller_search(graph, name)
  search_queue = graph[name].dup
  searched = {}

  loop do
    break if search_queue.empty?
    person = search_queue.shift
    next if searched[person]

    return [true, person] if person_is_seller?(person)
    search_queue += graph[person]
    searched[person] = true
  end
  [false, nil]
end
